### ISSUE TITLE
[bitnami/thanos] existingObjstoreSecret causes reboot loop of storegateway

### ISSUE THREAD

**Issue by pduchnovsky** [bitnami/charts#16046]
### Name and Version

bitnami/thanos 12.4.2

### What architecture are you using?

None

### What steps will reproduce the bug?

Deploy thanos chart with `existingObjstoreSecret` and storegateway enabled, values below

Secret set up using terraform as follows:
```
resource "kubernetes_secret" "thanos-objectstorage" {
  metadata {
    name      = "thanos-objectstorage"
    namespace = kubernetes_namespace.monitoring_namespace.metadata[0].name
  }

  data = {
    "objstore.yml" = <<EOF
type: AZURE
config:
  storage_account: "${azurerm_storage_account.thanos-storage.name}"
  storage_account_key: "${azurerm_storage_account.thanos-storage.primary_access_key}"
  container: "${azurerm_storage_container.thanos-backend.name}"
  endpoint: "blob.core.windows.net"
  max_retries: 10
EOF
  }

  type = "Opaque"
}
```

### Are you using any custom parameters or values?

```
# Custom values for thanos.
existingObjstoreSecret: thanos-objectstorage

storegateway:
  enabled: true
  grpc:
    server:
      tls:
        enabled: true
        autoGenerated: true
```

Note that other services we use (query, queryfrontend, compactor) seem to work fine with this setup, except storegateway

### What do you see instead?

From storegateway pod logs:
```
level=info ts=2023-04-13T12:14:42.307826482Z caller=factory.go:52 msg="loading bucket configuration"
level=error ts=2023-04-13T12:14:42.372893819Z caller=main.go:135 err="open /conf/index-cache.yml: no such file or directory
loading file /conf/index-cache.yml for index-cache.config-file
github.com/efficientgo/tools/extkingpin.(*PathOrContent).Content
    /go/pkg/mod/github.com/efficientgo/tools/extkingpin@v0.0.0-20220817170617-6c25e3b627dd/pathorcontent.go:79
main.runStore
    /app/cmd/thanos/store.go:304
main.registerStore.func1
    /app/cmd/thanos/store.go:214
main.main
    /app/cmd/thanos/main.go:133
runtime.main
    /usr/local/go/src/runtime/proc.go:250
runtime.goexit
    /usr/local/go/src/runtime/asm_amd64.s:1594
get content of index cache configuration
main.runStore
    /app/cmd/thanos/store.go:306
main.registerStore.func1
    /app/cmd/thanos/store.go:214
main.main
    /app/cmd/thanos/main.go:133
runtime.main
    /usr/local/go/src/runtime/proc.go:250
runtime.goexit
    /usr/local/go/src/runtime/asm_amd64.s:1594
preparing store command failed
main.main
    /app/cmd/thanos/main.go:135
runtime.main
    /usr/local/go/src/runtime/proc.go:250
runtime.goexit
    /usr/local/go/src/runtime/asm_amd64.s:1594"
```

**Comment by jotamartos** [comment]
Hi @pduchnovsky,

I was trying to reproduce the issue and obtain more information but all tests worked fine for me. In my case, I did not deployed the solution but rendered all the files using the template command. I created a test-values.yaml file with the values you provided

```
$ cat test-values.yaml
existingObjstoreSecret: thanos-objectstorage

storegateway:
  enabled: true
  grpc:
    server:
      tls:
        enabled: true
        autoGenerated: true
```

And rendered all files using the template command

```
$ helm template test -f test-values.yaml . | grep index-cache
```

>Note: I first removed the templates/NOTES.txt file as it generated errors

The previous command didn't return any index-cache line. That configuration parameter is inside an if block and it's not used unless indexCacheConfig is set (templates/storegateway/statefulset.yaml file)

```
            {{- if .Values.indexCacheConfig }}
            - --index-cache.config-file=/conf/index-cache.yml
            {{- end }}
```

Are you using extra parameters in your values file or edited the indexCacheConfig parameter in the default values.yaml file?

**Comment by pduchnovsky** [comment]
Oh yes we do use indexCacheConfig, forgot to paste it in initial post

Full yaml:
```
existingObjstoreSecret: thanos-objectstorage

indexCacheConfig:
  type: MEMCACHED
  config:
    addresses:
    - <added by terraform>
    auto_discovery: false
    dns_provider_update_interval: 60s
    max_async_buffer_size: 2
    max_async_concurrency: 2
    max_get_multi_batch_size: 0
    max_get_multi_concurrency: 0
    max_idle_connections: 2
    max_item_size: 0
    timeout: 0s

query:
  enabled: true
  stores: <added by terraform>
  grpc:
    client:
      tls:
        enabled: true
        autoGenerated: true

storegateway:
  enabled: true
  grpc:
    server:
      tls:
        enabled: true
        autoGenerated: true

compactor:
  enabled: true
```

When using `objstoreConfig` instead of `existingObjstoreSecret` it works fine..
```
objstoreConfig:
  type: AZURE
  config:
    storage_account: <added by terraform>
    storage_account_key: <added by terraform>
    container: <added by terraform>
    endpoint: "blob.core.windows.net"
    max_retries: 10
```

**Comment by jotamartos** [comment]
Hi @pduchnovsky,

The indexCacheConfig and objstoreConfig parameters are expecting strings when you set them in the values file. However, you are setting them like being objects and that's probably the reason why it's not working. You can take a look at this related support ticket where we tested that configuration 

https://github.com/bitnami/charts/issues/14362#issuecomment-1432838958

Let us know if that works for you.

**Comment by pduchnovsky** [comment]
> Hi @pduchnovsky,
> 
> The indexCacheConfig and objstoreConfig parameters are expecting strings when you set them in the values file. However, you are setting them like being objects and that's probably the reason why it's not working. You can take a look at this related support ticket where we tested that configuration
> 
> [#14362 (comment)](https://github.com/bitnami/charts/issues/14362#issuecomment-1432838958)
> 
> Let us know if that works for you.

As per above, values.yml contains this:
existingObjstoreSecret: thanos-objectstorage

The secret is set up as this:
```
    "objstore.yml" = <<EOF
type: AZURE
config:
  storage_account: "${azurerm_storage_account.thanos-storage.name}"
  storage_account_key: "${azurerm_storage_account.thanos-storage.primary_access_key}"
  container: "${azurerm_storage_container.thanos-backend.name}"
  endpoint: "blob.core.windows.net"
  max_retries: 10
EOF
```

There is no change in configuration for indexCacheConfig, just change from objstoreConfig to existingObjstoreSecret.

**Comment by jotamartos** [comment]
Hi @pduchnovsky,

That's a separate issue, right? We were talking about the index-cache.yml first and the objstore.yml file now. If you are using the `objstoreConfig` parameter, a secret should be created according to this template 

https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/objstore-secret.yaml

>Note: If you provide a valid ".Values.indexCacheConfig" value, the generated secret will also include that index-cache.yml file.

and the container should mount the secret and create the files properly 

https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/storegateway/statefulset-sharded.yaml#L284

If you see something wrong, you can always contribute by following [this guide](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md). 

Thanks

**Comment by pduchnovsky** [comment]
Let's simplify this, when we use this:

```
objstoreConfig:
  type: AZURE
  config:
    storage_account: <added by terraform>
    storage_account_key: <added by terraform>
    container: <added by terraform>
    endpoint: "blob.core.windows.net"
    max_retries: 10

indexCacheConfig:
  type: MEMCACHED
  config:
    addresses:
    - <added by terraform>
    auto_discovery: false
    dns_provider_update_interval: 60s
    max_async_buffer_size: 2
    max_async_concurrency: 2
    max_get_multi_batch_size: 0
    max_get_multi_concurrency: 0
    max_idle_connections: 2
    max_item_size: 0
    timeout: 0s

query:
  enabled: true
  stores: <added by terraform>
  grpc:
    client:
      tls:
        enabled: true
        autoGenerated: true

storegateway:
  enabled: true
  grpc:
    server:
      tls:
        enabled: true
        autoGenerated: true

compactor:
  enabled: true

queryFrontend:
  enabled: true
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: nginx
      xxxxx-contact: "<redacted>"
      cert-manager.io/cluster-issuer: letsencrypt-prod
      nginx.ingress.kubernetes.io/auth-response-headers: xxxxx
      nginx.ingress.kubernetes.io/auth-signin: xxxxx
      nginx.ingress.kubernetes.io/auth-url: xxxxx
      nginx.ingress.kubernetes.io/configuration-snippet: |
        xxxxx
    path: /
    hosts:
      - <added by terraform>
    tls: true
```

It works.

But when we remove objstoreConfig config and instead put `existingObjstoreSecret: thanos-objectstorage` it breaks and throws an error from first post.

**Comment by jotamartos** [comment]
TLDR: This line needs to be updated in the template

https://github.com/bitnami/charts/blob/main/bitnami/thanos/templates/storegateway/statefulset.yaml#L130

Long answer:

In this case, if you are manually creating a objstore secret, it needs to contain the info regarding the objstore.yml and index-cache.yml files and you need to set the `existingObjstoreSecret` parameter. 

Once you do that, the files are mounted in the container but the index-cache flag is not used because the .Values. indexCacheConfig parameter is not set. 

The error also happens the other way around, if you create a secret with just the objstore.yml file, that's the only file it's mounted in the container no matter if you set the .Values.indexCacheConfig parameter. 

If the mentioned line is updated to something similar to this one, your scenario should work. Can you try it?

```
            {{- if or .Values.indexCacheConfig ( and .Values.existingObjstoreSecret ( index ( lookup "v1" "Secret" .Release.Namespace .Values.existingObjstoreSecret ).data "index-cache.yml" )) }}
```

>Note: remember to include the objstore.yml and index-cache.yml information in the same secret.

I'm going to create a PR to solve this issue and you can also try with the new changes.

Thanks

**Comment by jotamartos** [comment]
PR created, can you take a look at it and confirm it works for you?

**Comment by pduchnovsky** [comment]
I have no option to test it before it's merged so I will test if afterwards.

**Comment by jotamartos** [comment]
Why are you saying that you can't test it? You can clone the repo, move to the PR's branch

```
git checkout bitnami/charts/issues/16046
```

and install the chart with the new changes 

```
helm repo update
cd bitnami/thanos
helm dependency build
helm install NAME -f TEST_VALUES.yml .
```

**Comment by pduchnovsky** [comment]
We deploy the env fully via terraform, not locally, and there is no option how to specify PR for deployment.
Unless you can provide full URL path to .tgz
https://registry.terraform.io/providers/hashicorp/helm/latest/docs/resources/release

**Comment by jotamartos** [comment]
You can always fork the repo and make the PR's branch (bitnami/charts/issues/16046) the default one as I did here 

https://github.com/jotamartos/charts

You can use that repo to test the changes if you want. It'll be really helpful.
